### ISSUE TITLE
Upgraded Django from 1.11 to 2.2 and now getting FileNotFoundError: [Errno 2] No such file or directory

### ISSUE THREAD

**Issue by matthewcornell** [django-compressor/django-compressor#1012]
Hi Folks. We are upgrading our app from Django 1.11 to 2.2, and we are now getting an intermittent "FileNotFoundError: [Errno 2] No such file or directory" when running our tests. We are using django-compressor version 2.4. (This part of the app was done by a former student, so please bear with me as I learn!) The error looks like this:

    FileNotFoundError: [Errno 2] No such file or directory: '/Users/cornell/IdeaProjects/forecast-repository/forecast_repo/staticfiles/CACHE/css/custom.94663ca72b05.acc6a25ed513.css'

However, that file *does* exist before and after running.

I've attached the stacktrace, but the trouble is at the `os.utime()` call here:

    compressor.storage.GzipCompressorFileStorage.save():
	    os.utime(orig_path, (stamp, stamp))

The only call to compress is in ouro base.html file that all files use:

    {% load static %}
    {% load compress %}
    {% compress css %}
        <link type="text/x-scss" rel="stylesheet" href="{% static 'css/custom.scss' %}">
    {% endcompress %}

The Django setup file base.py (which local, staging, and production use) has this configuration:

    INSTALLED_APPS = [
	...
	'compressor']

    COMPRESS_PRECOMPILERS = (
	('text/x-scss', 'django_libsass.SassCompiler'),  # https://github.com/torchbox/django-libsass -> {% compress css %}
    )

    STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'

    STATICFILES_FINDERS = [
	'django.contrib.staticfiles.finders.FileSystemFinder',
	'django.contrib.staticfiles.finders.AppDirectoriesFinder',
	'compressor.finders.CompressorFinder'  # Django-Compressor
    ]

    COMPRESS_STORAGE = 'compressor.storage.GzipCompressorFileStorage'
    # COMPRESS_ENABLED = True  # must enable this to use with Whitenoise. default: the opposite of DEBUG
    COMPRESS_OFFLINE = True  # "". default: False

    LIBSASS_OUTPUT_STYLE = 'compressed'  # default: 'nested'

I'm at my wits' end - any help would be really appreciated!

# stacktrace
    ======================================================================
    ERROR: test_url_get_general (forecast_app.tests.test_views_and_rest_api.ViewsTestCase)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/mock.py", line 1179, in patched
	return func(*args, **keywargs)
      File "/Users/cornell/IdeaProjects/forecast-repository/forecast_app/tests/test_views_and_rest_api.py", line 252, in test_url_get_general
	response = self.client.get(url, data={'unit': '', 'target': ''})
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/rest_framework/test.py", line 286, in get
	response = super().get(path, data=data, **extra)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/rest_framework/test.py", line 203, in get
	return self.generic('GET', path, **r)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/rest_framework/test.py", line 232, in generic
	method, path, data, content_type, secure, **extra)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/test/client.py", line 422, in generic
	return self.request(**r)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/rest_framework/test.py", line 283, in request
	return super().request(**kwargs)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/rest_framework/test.py", line 235, in request
	request = super().request(**kwargs)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/test/client.py", line 503, in request
	raise exc_value
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
	response = get_response(request)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
	response = self.process_exception_by_middleware(e, request)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
	response = wrapped_callback(request, *callback_args, **callback_kwargs)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/views/generic/base.py", line 71, in view
	return self.dispatch(request, *args, **kwargs)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/contrib/auth/mixins.py", line 108, in dispatch
	return self.handle_no_permission()
      File "/Users/cornell/IdeaProjects/forecast-repository/forecast_app/views.py", line 1007, in handle_no_permission
	return HttpResponseForbidden(render(self.request, '403.html').content)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/shortcuts.py", line 36, in render
	content = loader.render_to_string(template_name, context, request, using=using)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/loader.py", line 62, in render_to_string
	return template.render(context, request)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/backends/django.py", line 61, in render
	return self.template.render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/base.py", line 171, in render
	return self._render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/test/utils.py", line 96, in instrumented_test_render
	return self.nodelist.render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/base.py", line 937, in render
	bit = node.render_annotated(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/base.py", line 904, in render_annotated
	return self.render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/loader_tags.py", line 150, in render
	return compiled_parent._render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/test/utils.py", line 96, in instrumented_test_render
	return self.nodelist.render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/base.py", line 937, in render
	bit = node.render_annotated(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/django/template/base.py", line 904, in render_annotated
	return self.render(context)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/templatetags/compress.py", line 143, in render
	return self.render_compressed(context, self.kind, self.mode, forced=forced)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/templatetags/compress.py", line 119, in render_compressed
	rendered_output = compressor.output(mode, forced=forced, basename=file_basename)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/css.py", line 46, in output
	ret.append(subnode.output(*args, **kwargs))
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/css.py", line 48, in output
	return super(CssCompressor, self).output(*args, **kwargs)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/base.py", line 314, in output
	output = '\n'.join(self.filter_input(forced))
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/base.py", line 254, in filter_input
	for hunk in self.hunks(forced):
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/base.py", line 237, in hunks
	basename=basename)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/base.py", line 329, in handle_output
	return output_func(mode, content, forced, basename)
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/base.py", line 341, in output_file
	self.storage.save(new_filepath, ContentFile(content.encode(self.charset)))
      File "/Users/cornell/.local/share/virtualenvs/forecast-repository-aIKRoinE/lib/python3.6/site-packages/compressor/storage.py", line 72, in save
	with open(orig_path, 'rb') as f_in, open(compressed_path, 'wb') as f_out:
    FileNotFoundError: [Errno 2] No such file or directory: '/Users/cornell/IdeaProjects/forecast-repository/forecast_repo/staticfiles/CACHE/css/custom.94663ca72b05.acc6a25ed513.css'

    ----------------------------------------------------------------------
    Ran 1 test in 7.096s

    FAILED (errors=1)
    Destroying test database for alias 'default' ('test_forecast_repo')...

    Process finished with exit code 1

**Comment by matthewcornell** [comment]
Weird: sometimes it happens at this line:

            with open(orig_path, 'rb') as f_in, open(compressed_path, 'wb') as f_out:

**Comment by karyon** [comment]
have you been able to find the problem?

**Comment by mherrmann** [comment]
I'm experiencing this as well. For me, the problem started to appear when I started running Django's tests with the `--parallel` flag. My suspicion is that it's a multithreading issue.

```
Traceback (most recent call last):
  File ".../venv/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File ".../venv/lib/python3.9/site-packages/django/core/handlers/base.py", line 204, in _get_response
    response = response.render()
  File ".../venv/lib/python3.9/site-packages/django/template/response.py", line 105, in render
    self.content = self.rendered_content
  File ".../venv/lib/python3.9/site-packages/django/template/response.py", line 83, in rendered_content
    return template.render(context, self._request)
  File ".../venv/lib/python3.9/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 170, in render
    return self._render(context)
  File ".../venv/lib/python3.9/site-packages/django/test/utils.py", line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File ".../venv/lib/python3.9/site-packages/django/test/utils.py", line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File ".../venv/lib/python3.9/site-packages/django/test/utils.py", line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File ".../venv/lib/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File ".../venv/lib/python3.9/site-packages/compressor/templatetags/compress.py", line 143, in render
    return self.render_compressed(context, self.kind, self.mode, forced=forced)
  File ".../venv/lib/python3.9/site-packages/compressor/templatetags/compress.py", line 119, in render_compressed
    rendered_output = compressor.output(mode, forced=forced, basename=file_basename)
  File ".../venv/lib/python3.9/site-packages/compressor/css.py", line 46, in output
    ret.append(subnode.output(*args, **kwargs))
  File ".../venv/lib/python3.9/site-packages/compressor/css.py", line 48, in output
    return super(CssCompressor, self).output(*args, **kwargs)
  File ".../venv/lib/python3.9/site-packages/compressor/base.py", line 314, in output
    output = '\n'.join(self.filter_input(forced))
  File ".../venv/lib/python3.9/site-packages/compressor/base.py", line 254, in filter_input
    for hunk in self.hunks(forced):
  File ".../venv/lib/python3.9/site-packages/compressor/base.py", line 236, in hunks
    yield self.handle_output(kind, value, forced=True,
  File ".../venv/lib/python3.9/site-packages/compressor/base.py", line 329, in handle_output
    return output_func(mode, content, forced, basename)
  File ".../venv/lib/python3.9/site-packages/compressor/base.py", line 341, in output_file
    self.storage.save(new_filepath, ContentFile(content.encode(self.charset)))
  File ".../venv/lib/python3.9/site-packages/django/core/files/storage.py", line 54, in save
    return self._save(name, content)
  File ".../venv/lib/python3.9/site-packages/django/core/files/storage.py", line 298, in _save
    os.chmod(full_path, self.file_permissions_mode)
FileNotFoundError: [Errno 2] No such file or directory: '.../static-collected/CACHE/css/admin.d3124b542fdb.css'
```

**Comment by diox** [comment]
Hi, sorry for the huge delay in replying. https://github.com/django-compressor/django-compressor/commit/ef45175ecc9d5a649b112fdc156838aaff74bf72, which was part of the 3.0 release that I just pushed, might have fixed this. Can you check ? Thanks!

**Comment by diox** [comment]
This should be fixed by 00910e0bd5295044befacdad7b88249ec7bbf142. Please test the develop branch, and re-open if it's still a problem.

**Comment by diox** [comment]
This should be fixed in 4.0.
